controller/konnect/reconciler_kongplugin.go

```diff
@@ -88,130 +103,124 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, err
 	}
 
-	pluginBindingsToDelete := []configurationv1alpha1.KongPluginBinding{}
-	// pluginReferenceFound represents whether the plugin is referenced by any KongService
+	kongRouteList := configurationv1alpha1.KongRouteList{}
+	err = clientWithNamespace.List(ctx, &kongRouteList,
+		client.MatchingFields{
+			IndexFieldKongRouteOnReferencedPluginNames: kongPlugin.Namespace + "/" + kongPlugin.Name,
+		},
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	foreignRelations := ForeignRelations{
+		Service: lo.Filter(kongServiceList.Items,
+			func(s configurationv1alpha1.KongService, _ int) bool { return s.DeletionTimestamp.IsZero() },
+		),
+		Route: lo.Filter(kongRouteList.Items,
+			func(s configurationv1alpha1.KongRoute, _ int) bool { return s.DeletionTimestamp.IsZero() },
+		),
+		// TODO consumers and consumer groups
+	}
+	grouped, err := foreignRelations.GroupByControlPlane(ctx, r.client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// pluginReferenceFound represents whether the plugin is referenced by any resource.
 	var pluginReferenceFound bool
-	for _, kongService := range kongServiceList.Items {
-		if !kongService.DeletionTimestamp.IsZero() {
-			continue
-		}
+	groupedCombinations := grouped.GetCombinations()
 
-		var pluginSlice []string
+	// Delete the KongPluginBindings that are not used anymore.
+	if err := deleteUnusedKongPluginBindings(ctx, logger, clientWithNamespace, &kongPlugin, groupedCombinations, kongPluginBindingList.Items); err != nil {
+		return ctrl.Result{}, err
+	}
 
-		// get the referenced plugins from the KongService annotations
-		plugins, ok := kongService.Annotations[consts.PluginsAnnotationKey]
-		if !ok {
-			// if the konghq.com/plugins annotation is not present, we need to delete all the managed
-			// KongPluginBindings that reference the KongService
-			for _, pb := range bindingMapping.byServiceName[kongService.Name] {
-				if lo.ContainsBy(pb.OwnerReferences, func(ownerRef metav1.OwnerReference) bool {
-					if ownerRef.Kind == "KongPlugin" && ownerRef.Name == kongPlugin.Name && ownerRef.UID == kongPlugin.UID {
-						return true
-					}
-					return false
-				}) {
-					// The PluginBinding is dangling, so it needs to be deleted
-					pluginBindingsToDelete = append(pluginBindingsToDelete, pb)
-				} else {
-					pluginReferenceFound = true
-				}
+	for cpNN, relations := range groupedCombinations {
+		for _, rel := range relations {
+			pluginReferenceFound = true
+
+			builder := NewKongPluginBindingBuilder().
+				WithGenerateName(kongPlugin.Name + "-").
+				WithNamespace(kongPlugin.Namespace).
+				WithPluginRef(kongPlugin.Name).
+				WithControlPlaneRef(&configurationv1alpha1.ControlPlaneRef{
+					Type: configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef,
+					KonnectNamespacedRef: &configurationv1alpha1.KonnectNamespacedRef{
+						Namespace: cpNN.Namespace,
+						Name:      cpNN.Name,
+					},
+				})
+
+			l := kongPluginBindingList.Items
```

Comment:
Taking into account the volume of this function, naming this variable more verbose might be helpful for readers (I caught myself not understanding what `l` is in the switch below).
```suggestion
			kpbList := kongPluginBindingList.Items
```

Reply:
57ece90da2fd8801225fdc9e678179f837666926